### PR TITLE
make toUTF8 support up to 6 bytes

### PR DESCRIPTION
**Pull request by def-**
As noticed here: http://forum.nim-lang.org/t/897/2#5781

Not sure how to handle the case of an invalid Rune. Wouldn't it be nicer to restrict the rune to be in the valid range of Unicode characters?

### CODE REVIEW

**Comment by oderwat** [comment]
You mean this I guess: http://en.wikipedia.org/wiki/UTF-8#Invalid_code_points / in `read_code_point_from_utf8()`

**Comment by def-** [comment]
Mostly I meant negative numbers as well as anything higher than we can decode and encode anyway.

**Comment by oderwat** [comment]
For encoding I think that should throw an exception/crash (and "U+D800..U+DFFF" needs to be excluded, it is illegal).
For decoding it needs to do all the checks and should throw an exception too. Thats security related.

**Comment by Araq** [comment]
We used to raise an assertion failure in this case but it proved to be too hard to work with that. Garbage in, garbage out for strings works better. Yes. Really. It does.

**Comment by oderwat** [comment]
Not an assertion but an exception such that you can catch it. Maybe optional giving the possibility to read through garbage in a safe way. Afais there is no checking in the utf8 decoder at all anymore.

**Comment by Araq** [comment]
The type of the exception doesn't really matter, the problem is that there is no way to handle the exception. The best way to handle it is usually "hrm ok, just proceed, I'll look at the result" in my experience.

**Comment by oderwat** [comment]
But I want to reject / detect invalid UTF-8! So why can't this be a "try/expect" block and the function gets a parameter which says: catch it or ignore it. Probably also a parameter which contains the count of problems found. Some "read it fast and ignore if it kills kitten" variant could still exists. Maybe even with a "when" as "checked_utf8".

**Comment by Araq** [comment]
"Maybe throws depending on some parameter" doesn't work well with the effect system. We should simply have a 'validUtf8' proc that can return the index of the first wrong byte or something. Exceptions really suck here, as I usually want to continue with the rest of the string anyway, but "continue" is not the strength of exceptions. So that leaves us with replacing it with the general "invalid unicode" glyph. Well but that just throws away valuable information! So ... I see no alternative here to an additional proc that does sanity checking.

BTW: Aporia and Gedit refuse to open invalid files. That's very useful as I then need to fire up a less picky editor to edit the invalid UTF-8...

**Comment by Araq** [comment]
"make toUTF8 support up to 6 bytes". Why? 6 bytes are not valid UTF-8. Why did the former version of this code crash?

**Comment by def-** [comment]
> "make toUTF8 support up to 6 bytes". Why? 6 bytes are not valid UTF-8. Why did the former version of this code crash?

In the other direction 6 bytes are supported as well. But now I see that the limtation of UTF-8 to 4 bytes is actually newer than setting it to 6 bytes. I assumed this was the other way around. So, should I remove support for 6 bytes in `fastRuneAt` and others instead?

The former version crashed because of some invalid code I guess. In that case `chr(code)` was just called, which of course is out of the range of a `char`.

**Comment by oderwat** [comment]
Afaik: RFC 2044 & 2279 had 6 octets this was changed with RFC 3629 (in 2003) which defines 4 octets max and is whats now valid UTF-8.

Well, if something does not parse as valid UTF-8 it is not UTF-8. I see no room for interpretation here. 

You could "clean" the data into something which "may be close to what was supposed to be in the file" and for that you need to change it and this **needs to be notified** to the caller (IMHO).

What about adding a callback which gets buffer + offset and returns the replacement char? Letting this raise an exception could return the offset or just bail out of bad UTF-8. Wouldn't that be handled correctly by the effect system?

Probably this gives some ideas: http://utfcpp.sourceforge.net/

File: lib/pure/unicode.nim [hunk 1]
```diff
@@ -118,21 +118,35 @@ proc toUTF8*(c: Rune): string {.rtl, extern: "nuc$1".} =
...
+    result[0] = chr(i shr 30 or 0b1111110_0)
+    result[1] = chr(i shr 24 and ones(6) or 0b10_0000_00)
+    result[2] = chr(i shr 18 and ones(6) or 0b10_0000_00)
+    result[3] = chr(i shr 12 and ones(6) or 0b10_0000_00)
+    result[4] = chr(i shr 6 and ones(6) or 0b10_0000_00)
+    result[5] = chr(i and ones(6) or 0b10_0000_00)
   else:
-    result = newString(1)
-    result[0] = chr(i)
+    discard # error, exception?
```
Comment (def-) [inline lib/pure/unicode.nim]: I'm still not sure what to do in this case. Right now we return a null pointer. Is that ok?

**Pull request merged**